dbms/src/Common/HashTable/HashTable.h

```diff
@@ -763,6 +772,22 @@ class HashTable :
             emplaceNonZero(x, it, inserted, hash_value);
     }
 
+    void ALWAYS_INLINE emplace(Key x, iterator & it, bool & inserted, size_t hash_value, DB::Arena & pool)
+    {
+        if (!emplaceIfZero(x, it, inserted, hash_value))
+        {
+            emplaceNonZero(x, it, inserted, hash_value);
+            if constexpr (std::is_same_v<Key, StringRef>)
```

Comment:
So in the end the HashTable decides whether to put the key to arena or not, looking at its type? This looks convoluted. Can we decide this in the caller and just call the proper method? Or even better, just work with Arena in the caller, and don't make HashTable know about it at all.

Reply:
It's not trivial since StringHashMap has different submaps and only the string submap does allocation. Adding a new method will require multiple if constexpr hanging around

Reply:
You mean, because of multiplexing in dispatch(), the callers like HashMethodBase::emplaceKeyImpl() can't know whether they should put the key to Arena or not? This is a problem indeed. Still, I think HashTable is supposed to be a generic container, and so it shouldn't contain any logic specific to the particular use case with StringHashTable and Arena. Maybe we could make this decision in StringHashTable itself?

Or maybe we could move the Arena-related stuff even higher in the call graph. I'll try to find a solution next week, but this will take some time. The hashing-related code in CH seems rather convoluted and I'm having a hard time understanding it.

Reply:
@akuzm Gentle ping.

Reply:
Replied on Telegram.